### PR TITLE
fix(clients): make POST body related to the endpoint, assert body in tests

### PR DESCRIPTION
**Pull request by shortcuts**
## 🧭 What and Why

🎟 JIRA Ticket: https://algolia.atlassian.net/browse/DI-321

The issue was discovered in https://github.com/algolia/api-clients-automation/pull/823, but I've moved the fix for the clients here.

### Changes included:

The goal of this PR is to assert the body of each requests in the CTS even when we don't expect to have one.

On some `POST` requests, the body can be optional (e.g. `browse`), but needs to be a valid JSON object, when no body is expected at all (e.g. `clearAllSynonyms`), an empty string is considered as valid. ~In any case, an empty object (`'{}'`) is considered as valid.~ Some endpoints **do not expect a body at all**, so we need to move that logic at the method level since it can't be factorized.

### Issue

1. Calling the `browse` method with only an `indexName` throws an empty body error
2. Calling the `clearRules` method with a body throws an unexpected body error

### Next

Fix `null` body for `GET` and `DELETE` methods on the PHP client, once @damcou is back: https://algolia.atlassian.net/browse/APIC-591

## 🧪 Test

- CI
- Playground:
  - `deleteApiKey` for the `DELETE` method change
  - `browse` with empty object only for the `POST` method change

### CODE REVIEW

**Comment by shortcuts** [comment]
> Looks good ! Hopefully all endpoints accept that !

Doing more tests on the `DELETE` side before merging

File: templates/java/tests/requests/requests.mustache [hunk 1]
```diff
@@ -66,12 +66,14 @@ class {{client}}RequestsTests {
...
         assertEquals(req.path, "{{{request.path}}}");
         assertEquals(req.method, "{{{request.method}}}");
-
         {{#request.body}}
         assertDoesNotThrow(() -> {
             JSONAssert.assertEquals("{{#lambda.escapeQuotes}}{{{request.body}}}{{/lambda.escapeQuotes}}", req.body, JSONCompareMode.STRICT);
         });
         {{/request.body}}
+        {{^request.body}}
+        assertEquals(req.body, null);
```
Comment (millotp) [inline templates/java/tests/requests/requests.mustache:75]: you can use `assertNull` :)

**Review by millotp — COMMENTED** [review]
Looks good for java and php ! But you are still getting undefined for js

**Comment by shortcuts** [comment]
> Looks good for java and php ! But you are still getting undefined for js

Yup I'm importing the echoRequester fixes from the other PR, just making sure it works properly

File: templates/java/tests/requests/requests.mustache (hunk 1 above)
Comment (shortcuts) [reply templates/java/tests/requests/requests.mustache]: done!

File: clients/algoliasearch-client-javascript/packages/client-common/src/transporter/createTransporter.ts [hunk 2]
```diff
@@ -236,41 +240,23 @@ export function createTransporter({
...
-        ...baseRequestOptions.headers,
-      },
-    };
-
-    const isRead = request.useReadTransporter || request.method === 'GET';
+    /**
+     * A read request is either a `GET` request, or a request that we make
+     * via the `read` transporter (e.g. `search`).
+     */
+    const isRead = Boolean(
```
Comment (millotp) [inline clients/algoliasearch-client-javascript/packages/client-common/src/transporter/createTransporter.ts:250]: is `Boolean()` needed ? can it be anything other than a boolean ?

File: clients/algoliasearch-client-php/lib/RequestOptions/RequestOptionsFactory.php [hunk 3]
```diff
@@ -50,7 +50,8 @@ private function normalize($options)
             'headers' => [
                 'x-algolia-application-id' => $this->config->getAppId(),
                 'x-algolia-api-key' => $this->config->getAlgoliaApiKey(),
-                'User-Agent' => $this->config->getAlgoliaAgent() !== null
+                'User-Agent' =>
```
Comment (millotp) [inline clients/algoliasearch-client-php/lib/RequestOptions/RequestOptionsFactory.php:53]: I think the php changes can be moved to another PR until @damcou is back

File: tests/output/javascript/yarn.lock [hunk 4]
```diff
@@ -35,12 +35,6 @@ __metadata:
   languageName: node
```
Comment (millotp) [inline tests/output/javascript/yarn.lock:35]: weird that this wasn't generated by the CTS, maybe you can include this to the file to push from the CI ?

File: clients/algoliasearch-client-java-2/algoliasearch-core/src/main/java/com/algolia/utils/JSONBuilder.java [hunk 5]
```diff
@@ -23,6 +23,9 @@ public ObjectMapper build() {
     mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
     mapper.enable(SerializationFeature.WRITE_ENUMS_USING_TO_STRING);
     mapper.enable(DeserializationFeature.READ_ENUMS_USING_TO_STRING);
+
+    // Allow defaulting a POST body property to an empty Object instance
+    mapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
```
[the remark this answers is not in the document]
Comment (shortcuts) [reply clients/algoliasearch-client-java-2/algoliasearch-core/src/main/java/com/algolia/utils/JSONBuilder.java:28]: Yep that's the part I'm not really sure about, it's what they recommend when trying to provide an empty object but I'd rather about it.

It's needed for the `new Object()`, do you have an alternative in mind?

File: clients/algoliasearch-client-javascript/packages/client-common/src/transporter/createTransporter.ts (hunk 2 above)
Comment (shortcuts) [reply clients/algoliasearch-client-javascript/packages/client-common/src/transporter/createTransporter.ts:250]: Good catch it's a leftover!

File: clients/algoliasearch-client-java-2/algoliasearch-core/src/main/java/com/algolia/utils/JSONBuilder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (millotp) [reply clients/algoliasearch-client-java-2/algoliasearch-core/src/main/java/com/algolia/utils/JSONBuilder.java:28]: maybe have a special case for this in the `serialize` method, like `if(obj.getClass().getName().equals("Object")) return "{}";`

File: clients/algoliasearch-client-php/lib/RequestOptions/RequestOptionsFactory.php (hunk 3 above)
Comment (shortcuts) [reply clients/algoliasearch-client-php/lib/RequestOptions/RequestOptionsFactory.php:53]: The changes for the PHP client here still fixes the `POST` methods that were not working previously (e.g. `browse`). However, I've removed the transporter/CTS changes in the meantime

File: clients/algoliasearch-client-java-2/algoliasearch-core/src/main/java/com/algolia/utils/JSONBuilder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (shortcuts) [reply clients/algoliasearch-client-java-2/algoliasearch-core/src/main/java/com/algolia/utils/JSONBuilder.java]: seems to work here: https://github.com/algolia/api-clients-automation/pull/849/commits/8ee4c28f1ba792f31a171524a5d2c0f17777687d

[the remark this answers is not in the document]
Comment (millotp) [reply clients/algoliasearch-client-java-2/algoliasearch-core/src/main/java/com/algolia/utils/JSONBuilder.java]: very hacky but as least it's safer than the flag !

**Review by millotp — APPROVED** [review]
looks good !
